### PR TITLE
Verify peer

### PR DESCRIPTION
**Pull request by bester**
This patch adds an option to enable the curl verify_peer option. It works around the s3 virtual host certificate wildcard mismatch by using the top-level s3 hostname in the URI and then explicitly setting the Host header in the request.

### CODE REVIEW

File: src/request.c [hunk 1]
```diff
@@ -433,6 +437,22 @@ static S3Status compose_standard_headers(const RequestParams *params,
...
+    if (params->bucketContext.uriStyle == S3UriStyleVirtualHost) {
+        const char *requestHostName = params->bucketContext.hostName
+                ? params->bucketContext.hostName : defaultHostNameG;
+
+        size_t len = snprintf(values->hostHeader, sizeof(values->hostHeader), "Host: %s.%s",
+                        params->bucketContext.bucketName, requestHostName);
+        if (len >= sizeof(values->hostHeader)) {
+            return S3StatusUriTooLong;
+        }
+        while (is_blank(values->hostHeader[len])) {
```
Comment (bji) [inline src/request.c]: Shouldn't this be testing is_blank(values->hostHeader[len - 1] ???

The value at [len] will start out as the  zero-terminator that snprintf put there, so this while loop will never do anything.

**Comment by bji** [comment]
Additional question: how is the curl failure to make the request due to SSL certificate verification error distinguishable from other connection initialization errors?  If it's not, then I think it would be really good to detect this error specifically and provide another failure status code for it.  Because I can certainly see how an app would want to know whether the failure was due to cert issues vs. other issues.

Thanks for the patch by the way, if we can address just a few issues here I'd be happ to pull it.

**Comment by bji** [comment]
Ignore the comment at line 450.  github won't let me delete it.

**Comment by bester** [comment]
I added code to map CURLE_PEER_FAILED_VERIFICATION to the S3StatusServerFailedVerification error code. S3StatusServerFailedVerification has also been used to handle the CURLE_SSL_CACERT status as well. I'm not sure whether you'll want to be able to distinguish those or not.

I think the first is  host != uri name and the latter is certificate chain problems.

**Comment by bji** [comment]
Thank you, this is fine.  I see that I neglected to add documentation into libs3.h identifying when the "ConnectionFailed" and "FailedVerification" errors could be returned by S3_ functions.  I will fix the docs myself, not really fair to ask you to do that :)

**Pull request merged**